src/censored.jl

```diff
@@ -0,0 +1,257 @@
+"""
+    censored(d::UnivariateDistribution, l::Real, u::Real)
+
+Censor a univariate distribution `d` to the interval `[l, u]`.
+
+The density function (or mass function for discrete `d`) of the censored distribution is
+```math
+f(x; d, l, u) = \\begin{cases}
+    P_d(x \\le l), & x \\le l \\\\
+    P_d(x),         & l < x < u \\\\
+    P_d(x \\ge u), & x \\ge u \\\\
+  \\end{cases},
+```
+where ``P_d(x)`` is the density (or mass) function of `d`, and ``P_d(x \\le b)`` is the
+cumulative distribution function of ``d`` evaluated at ``x = b``.
+If ``x`` is a random variable from ``d``, then `clamp(x, l, u)` is a random variable from
+its censored version. Note that this implies that even if ``d`` is continuous, its censored
+form has discrete mixture components at the bounds.
+
+The lower bound `l` can be finite or `-Inf` and the upper bound `u` can be finite or
+`Inf`. The function throws an error if `l > u`.
+
+The function falls back to constructing a [`Censored`](@ref) wrapper.
+
+# Implementation
+
+To implement a specialized censored form for distributions of type `D`, the method
+`censored(d::D, l::T, u::T) where {T <: Real}` should be implemented.
+"""
+function censored(d::UnivariateDistribution, l::Real, u::Real)
+    return censored(d, promote(l, u)...)
+end
+censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real} = Censored(d, l, u)
+
+"""
+    Censored
+
+Generic wrapper for a censored distribution.
+"""
+struct Censored{D<:UnivariateDistribution, S<:ValueSupport, T <: Real} <: UnivariateDistribution{S}
+    uncensored::D      # the original distribution (uncensored)
+    lower::T      # lower bound
+    upper::T      # upper bound
+    function Censored(d::UnivariateDistribution, l::T, u::T) where {T <: Real}
+        l ≤ u || error("the lower bound must be less than or equal to the upper bound")
+        new{typeof(d), value_support(typeof(d)), T}(d, l, u)
+    end
+end
+Censored{D,S,T}(l::T, u::T, params...) where {D,S,T} = Censored(D(params...), l, u)
+
+function censored(d::Censored, l::T, u::T) where {T <: Real}
+    return Censored(d.uncensored, max(l, d.lower), min(u, d.upper))
+end
+
+params(d::Censored) = (d.lower, d.upper, params(d.uncensored)...)
+partype(d::Censored) = Base.promote_eltype(partype(d.uncensored), d.lower)
+Base.eltype(::Type{Censored{D,S,T}}) where {D,S,T} = Base.promote_eltype(eltype(D), T)
+
+#### Range and Support
+
+islowerbounded(d::Censored) = islowerbounded(d.uncensored) || isfinite(d.lower)
+isupperbounded(d::Censored) = isupperbounded(d.uncensored) || isfinite(d.upper)
+
+minimum(d::Censored) = max(minimum(d.uncensored), d.lower)
+maximum(d::Censored) = min(maximum(d.uncensored), d.upper)
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    return d.lower ≤ x ≤ d.upper && insupport(d.uncensored, x)
```

Comment:
If `x` is equal to one of the bounds, this will return `false` unless the bound is also in the support of `d`, but that doesn't need to be the case.

Reply:
Hmm I guess I misunderstood the definition of a censored distribution then: I assume it is always a mixture of the original distribution restricted to the censored interval (whose support may or may not include the lower and upper bound) and two Diracs at the lower and upper bound. I.e., the lower and upper bound would always be in the support of the censored distribution. Is this not the case?

Reply:
Yes, I'm leaning toward that the bounds should always be included. That's what the implementations currently all assume, except for this function. But this can cause some odd behavior for discrete distributions (see https://github.com/JuliaStats/Distributions.jl/pull/1470#discussion_r783502531).

I guess why you would want this is for cases where instead of replacing a censored measurement by its known minimum or maximum value, you set the bounds to something shifted by a small amount to distinguish it, which would only work in the discrete case anyways.

Reply:
The support of a distribution should always be a closed set so so nothing weird should happen (https://en.wikipedia.org/wiki/Support_(mathematics)) , but again our "private" use might disagree.
The only thing which could happen is if the censoring is outside the support and then nothing happens, e.g. in case of right censoring and having the left censoring bound at `-Inf`.

Reply:
> If `x` is equal to one of the bounds, this will return false unless the bound is also in the support of `d`, but that doesn't need to be the case.

So I think this is correct.

Reply:
> > If `x` is equal to one of the bounds, this will return false unless the bound is also in the support of `d`, but that doesn't need to be the case.
> 
> So I think this is correct.

To be sure I understand, what precisely do you mean by "this" here?

Reply:
I meant that
```
return d.lower ≤ x ≤ d.upper && insupport(d.uncensored, x)
```
is the correct definition, but I start to understand the issues, wait...

Reply:
This is a problem:
```

julia> insupport(Poisson(1), 1)
true

julia> insupport(Poisson(1), 1.5)
false

julia> insupport(Poisson(1), 2)
true

julia> clamp(rand(Poisson(1)), -Inf, 1.5)
1.5
```

Reply:
To be clear, my understanding of the definition of this distribution was
```julia
return (d.lower ≤ x ≤ d.upper && insupport(d.uncensored, x)) || (x == d.lower && d.lower > -Inf) || (x == d.upper && d.upper < Inf)
```
(still with the +-Inf convention if it is unbounded). I.e., one always has atoms at `d.lower` and `d.upper` if they are bounded.

Reply:
> This is a problem:

Exactly, therefore I assumed the definition should be as described in the last comment.

Reply:
I think almost, we should check `... || (x == d.lower && cdf(d, x) > 0) || ...` to determine if `x == d.lower` is in the support

Reply:
This seems even better and leads to correct results also for something like `censored(Uniform(0, 1), -1.5, 2.5)`.